File: src/co/edu/uptc/music/servlets/PlaylistsServlet.java
package co.edu.uptc.music.servlets;

import com.google.gson.Gson;

import java.io.IOException;
import java.io.PrintWriter;
import java.util.ArrayList;

import javax.servlet.ServletException;
import javax.servlet.annotation.WebServlet;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import co.edu.uptc.music.logic.managers.PlaylistsManager;
import co.edu.uptc.music.logic.managers.SongsManager;
import co.edu.uptc.music.logic.managers.UsersManager;
import co.edu.uptc.music.logic.models.Playlist;
import co.edu.uptc.music.logic.models.Song;
import co.edu.uptc.music.logic.models.User;
import co.edu.uptc.music.logic.models.UserType;

@WebServlet(name = "PlaylistsServlet", urlPatterns = {"/PlaylistsServlet"})
public class PlaylistsServlet extends HttpServlet {

    private UsersManager users = new UsersManager();
    private SongsManager songsManager = new SongsManager();
    private PlaylistsManager playlists = new PlaylistsManager();

    protected void processRequest(HttpServletRequest request, HttpServletResponse response)
            throws Exception {
        response.setContentType("application/json");
        response.setCharacterEncoding("UTF-8");
        Gson gson = new Gson();
        int opc = Integer.valueOf(request.getParameter("data"));
        users.load();
        try (PrintWriter writer = response.getWriter()) {
            String username = request.getParameter("username");
            User u = users.findItem(username);
            if (u != null) {
                if (opc == 1) {
                    String songId = request.getParameter("songId");
                    playlists.createFavsPlaylist(u.getId());
                    if (playlists.addSongsToPlaylist("F" + u.getId().substring(1), songId)) {
                        writer.print("{\"code\": 1}");
                    }
                } else if (opc == 2) {
                    songsManager.loadFavorites(u.getId());
                    ArrayList<Song> favorites = songsManager.getList();
                    if (favorites.size() > 0) {
                        writer.print("{\"songs\":" + gson.toJson(favorites) + "}");
                    }
                } else if (opc == 3) {
                    playlists.load(u.getId());
                    if (playlists.getList().size() > 0) {
                        writer.print("{\"lists\":" + gson.toJson(playlists.getList()) + "}");
                    }
                } else if (opc == 4) {
                    if (u.getType() == UserType.ADMIN || u.getType() == UserType.ARTIST) {
                        writer.print("{\"code\": 1}");
                    }
                } else if (opc == 5) {
                    if (playlists.createNewPlaylist(request.getParameter("listid"), u.getId())) {
                        writer.print("{\"code\": 1}");
                    }
                } else if (opc == 6) {
                    playlists.load(u.getId());
                    String listname = request.getParameter("listid");
                    Playlist target = null;
                    for (Playlist p : playlists.getList()) {
                        if (p.getId().equalsIgnoreCase(listname)) target = p;
                    }
                    if (target != null) {
                        playlists.addSongsToPlaylist(target.getId(),
                                request.getParameter("songid"));
                        writer.print("{\"code\": 1}");
                    }
                } else if (opc == 7) {
                    String listname = request.getParameter("listid");
                    songsManager.loadSongsFromPlaylists(u.getId(), listname);
                    ArrayList<Song> songsFromPlaylist = songsManager.getList();
                    if (songsFromPlaylist.size() > 0) {
                        writer.print("{\"songs\":" + gson.toJson(songsFromPlaylist) + "}");
                    }
                }
            }
            writer.close();
        } catch (Exception ignored) {
        }
    }


    @Override
    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws
            ServletException, IOException {
        try {
            processRequest(req, resp);
        } catch (Exception ignored) {
        }
    }

    @Override
    protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws
            ServletException, IOException {
        try {
            processRequest(req, resp);
        } catch (Exception ignored) {
        }
    }
}

File: src/co/soundwave/servlets/PlaylistsServlet.kt
package co.soundwave.servlets

import co.soundwave.extensions.ignore
import co.soundwave.managers.PlaylistsManager
import co.soundwave.managers.SongsManager
import co.soundwave.managers.UsersManager
import com.google.gson.Gson
import javax.servlet.annotation.WebServlet
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@WebServlet(name = "PlaylistsServlet", urlPatterns = ["/PlaylistsServlet"])
class PlaylistsServlet : BaseServlet() {
    
    private val users = UsersManager()
    private val songsManager = SongsManager()
    private val playlists = PlaylistsManager()
    
    override fun processRequest(
        request: HttpServletRequest,
        response: HttpServletResponse,
        type: Type
                               ) {
        response.contentType = "application/json"
        response.characterEncoding = "UTF-8"
        val gson = Gson()
        val opc = Integer.valueOf(request.getParameter("data"))
        users.load()
        ignore {
            response.writer.use { writer ->
                val username = request.getParameter("username")
                val u = users.findItem(username)
                if (u != null) {
                    playlists.load(u.id)
                    if (opc == 1) {
                        val songId = Integer.parseInt(request.getParameter("songId"))
                        playlists.createFavsPlaylist(u.id)
                        if (playlists.addSongToFavorites(songId)) {
                            writer.print("{\"code\": 1}")
                        }
                    } else if (opc == 2) {
                        val favs = playlists.getList().firstOrNull { it.name == "Favorites" }
                        favs?.let {
                            writer.print(
                                "{\"songs\":" + gson.toJson(
                                    songsManager.getSongsInPlaylist(it.id)) + "}")
                        } ?: {
                            writer.print(
                                "{\"songs\": []}")
                        }()
                    } else if (opc == 3) {
                        if (playlists.getList().isNotEmpty()) {
                            writer.print("{\"lists\":" + gson.toJson(playlists.getList()) + "}")
                        }
                    } else if (opc == 4) {
                        println("Opción 4 ... kejesto?")
                        /* TODO: Check
                    if (u.getType() == UserType.ADMIN || u.getType() == UserType.ARTIST) {
                        writer.print("{\"code\": 1}");
                    }
                    */
                    } else if (opc == 5) {
                        if (playlists.createNewPlaylist(request.getParameter("listid"), u.id)) {
                            writer.print("{\"code\": 1}")
                        }
                    } else if (opc == 6) {
                        val songId = Integer.parseInt(request.getParameter("songid"))
                        val list = playlists.findItem(request.getParameter("listid"))
                        if (list != null) {
                            playlists.addSongToPlaylist(list.id, songId)
                            writer.print("{\"code\": 1}")
                        }
                    } else if (opc == 7) {
                        val list =
                            playlists.findItem(Integer.parseInt(request.getParameter("listid")))
                        if (list != null) {
                            val songsFromPlaylist = songsManager.getSongsInPlaylist(list.id)
                            if (songsFromPlaylist.size > 0) {
                                writer.print("{\"songs\":" + gson.toJson(songsFromPlaylist) + "}")
                            }
                        }
                    } else if (opc == 8) {
                        val list =
                            playlists.findItem(Integer.parseInt(request.getParameter("listid")))
                        if (list != null && playlists.delete(list.id)) {
                            writer.print("{\"code\": 1}")
                        }
                    }
                }
                writer.close()
            }
        }
    }
    
    override fun getServletInfo(): String = "Playlists Servlet"
}